File: Algorithms/16. 3 Sum Closest/3 Sum Closest.java
//Author: Tushar Jaiswal
//Creation Date: 10/03/2015

/*Given an array S of n integers, find three integers in S such that the sum is closest to a given number, target. Return the sum of the three integers. You may assume that each input would have exactly one solution.
Input: 
S=[0,0,0] & target=1
Output: 0
S=[-1,2,1,-4] & target=1
Output: 2
S=[-3,0,1,2] & target=1
Output: 0
*/
public class Solution {
    public int threeSumClosest(int[] nums, int target) {
        int closestSum=0;
        Arrays.sort(nums);
        
        for(int i=0; i<nums.length-2; i++)
        {
            int j = i+1;
            int k = nums.length-1;
            int optimalTwoSum = target-nums[i];
            
            if(i == 0)
            {
                closestSum = nums[i] + nums[j] + nums[k];
            }
            
            if(i > 0 && nums[i] == nums[i-1])//Avoid duplicate values
            { continue; }
            
            while(j < k)
            {
                int twoSum = nums[j] + nums[k];
                if(twoSum == optimalTwoSum)
                { return target;}
                else if(Math.abs(optimalTwoSum - twoSum) < Math.abs(target - closestSum))
                {
                    closestSum = nums[i] + twoSum;
                }
                if(twoSum < optimalTwoSum)
                { j++; }
                else 
                if(twoSum > optimalTwoSum)
                { k--; }
            }
        }
        return closestSum;
    }
}

File: Algorithms/16. 3 Sum Closest/3SumClosest.cs
//Author: Tushar Jaiswal
//Creation Date: 02/08/2019

/*Given an array nums of n integers and an integer target, find three integers in nums such that the sum is closest to target. Return the sum of the three integers. You may assume that each input would have exactly one solution.

Example:
Given array nums = [-1, 2, 1, -4], and target = 1.
The sum that is closest to the target is 2. (-1 + 2 + 1 = 2).*/

public class Solution {
    public int ThreeSumClosest(int[] nums, int target) {
        Array.Sort(nums);
        int closest = nums[0] + nums[1] + nums[2];

        for(int i = 0; i < nums.Length - 2; i++)
        {
            int left = i + 1;
            int right = nums.Length - 1;

            if(i > 0 && nums[i - 1] == nums[i])
            { continue; }

            while(left < right)
            {
                int sum = nums[i] + nums[left] + nums[right];
                if(sum == target)
                {
                    return sum;
                }
                if(Math.Abs(target - sum) < Math.Abs(target - closest))
                {
                    closest = sum;
                }
                if(sum < target)
                {
                    left++;
                }
                else
                {
                    right--;
                }
            }
        }
        return closest;
    }
}
